sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedPlainValuesReader.java

```diff
@@ -63,115 +58,139 @@ public final void readBooleans(int total, WritableColumnVector c, int rowId) {
     }
   }
 
+  private ByteBuffer getBuffer(int length) {
+    try {
+      return in.slice(length).order(ByteOrder.LITTLE_ENDIAN);
+    } catch (IOException e) {
+      throw new ParquetDecodingException("Failed to read " + length + " bytes", e);
+    }
+  }
+
   @Override
   public final void readIntegers(int total, WritableColumnVector c, int rowId) {
-    c.putIntsLittleEndian(rowId, total, buffer, offset - Platform.BYTE_ARRAY_OFFSET);
-    offset += 4 * total;
+    int requiredBytes = total * 4;
+    ByteBuffer buffer = getBuffer(requiredBytes);
+
+    for (int i = 0; i < total; i += 1) {
```

Comment:
It looks that way, but it actually replaces a similar loop: https://github.com/apache/spark/blob/master/sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/OnHeapColumnVector.java#L283-L291

The main problem is that ByteBufffer isn't supported in the column vectors. That seems beyond the scope of this PR.

Reply:
Couldn't it also be writing to an `OffHeapColumnVector`? https://github.com/apache/spark/blob/master/sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java#L199

If so, I think the copy is 1MB at a time: https://github.com/apache/spark/blob/master/common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java#L189

I agree that ByteBuffer shouldn't be supported in this PR. But there's an opportunity to use the bulk copy APIs which would benefit from any future optimization that happens. Plus even if the copy does eventually become a loop inside the column vector implementation, there's more chance of the JIT unrolling the loop since it's smaller.

Reply:
I think that better integration between `ByteBuffer` and `ColumnVector` would be addressed in another PR.
Since tableCache also uses `ByteBuffer`, it would be good to discuss in another PR.
cc: @cloud-fan

Reply:
Agreed that fixing the `ByteBuffer` / `ColumnVector` interaction should be dealt with elsewhere. I'm just raising the possibility of _regressing_ the read path here because the copies are less efficient. Since it's going to be a while before 2.4.0, that might be ok if we commit to fixing it - but it superficially seems like a manageable change to the PR since the code to call the bulk APIs is already there.  What do you think?

Reply:
The reason why I moved the loop out was that I didn't think that using the byte[] API would actually be better. Parquet doesn't guarantee that these byte buffers are on the heap and backed by byte arrays, so we would need to copy the bytes out of the buffer into an array only to copy them again in the column vector call. Two copies (and possibly allocation) seemed worse than moving the loop.

We could catch the case where the buffers are on-heap and make the bulk call. The drawback is that this will be temporary and will be removed when ColumnVector supports ByteBuffer. And, it only works/matters when Parquet uses on-heap buffers and Spark uses off-heap buffers. Is that worth the change to this PR? I can take a shot at it if you think it is. I'd rather update ColumnVector and then follow up though.

Reply:
Isn't that what `hasArray()` is for though? If the buffers are backed by a byte array, `hasArray()` returns true and accessing the byte array via `array()` should be 0 cost. (If `array()` actually copies any data, that would invalidate this line of reasoning but would also be unexpected).

So for example, here you'd have:
    
    public final void readIntegers(int total, WritableColumnVector c, int rowId) {
        int requiredBytes = total * 4;
        ByteBuffer buffer = getBuffer(requiredBytes);
        if (buffer.hasArray()) {
          c.putIntsLittleEndian(rowId, total, buffer.array(), 0);         
        } else {
            for (int i = 0; i < total; i += 1) {
                c.putInt(rowId + i, buffer.getInt());
            }
        }
    }

This seems to be the same pattern that's in `readBinary()`, below. Let me know if I'm missing something!

Reply:
Yeah, exactly. We can detect that the buffer is backed by an array and use the other call. If we think this is worth it as a short-term fix, I'll update this PR.

Reply:
Updated.